## TITANIC

## importing necessary libraries

In [433]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
% matplotlib inline


## getting data from kaggle

In [378]:
url1='https://storage.googleapis.com/kaggle-competitions-data/kaggle/3136/test.csv?GoogleAccessId=web-data@kaggle-161607.iam.gserviceaccount.com&Expires=1528645563&Signature=oVei%2FNwjQJDUkZ%2FSXJc6O8bB7R4%2FI1ty5zyDx3pX53nJM0owVwPb%2BBtCn0M7f%2BijYW15OWiMRkwDzYZPaFjTxnWD10GztDjuBZVey8d3tPFcVQwNoZZC4DY0GkTc4i8CWFfmNgwkkECb%2Bc4%2Fk4SFsFs3voYbLbuxWUUZorn%2FMukHZoahKfWcO6HlRxOEdK1cQOM3bXtHk9mjyukzDxLtcD4GOzm33ZHt9SSzNlKeg4JIuJcHQm06Q6oEjasKPnrN9Y1oWvesE9xN32ul0QbuHfE6f7XIdL127Rypplxb2E9ihDq7l%2B%2B4K5RhiYhpL%2FHxidivpWkFhxwQtecRaVGiiQ%3D%3D'
test=pd.read_csv(url1)

url='https://storage.googleapis.com/kaggle-competitions-data/kaggle/3136/train.csv?GoogleAccessId=web-data@kaggle-161607.iam.gserviceaccount.com&Expires=1528645754&Signature=Id6Ls3UELTm%2FvOuesVykhSLNSLJNsg2k7C%2BLV9j86SoybSCuPA5fnzE7SEPEOqw4Oq7UT5rB16HfaBcKC4WlMx0ahOyftHWmB7MO5sI5swHFdDmkQOzOfeStUPwh9qhZoe8ZSbyOV1T%2BU2YK6Jd4YO3f7W7G1PXcQdViRLSIQzAy18NU00JyUFlStegHUfiQGzYP1pESGP%2FBFZ9mfz4gVIm%2BoOjk9t6nzSuImdi7c1p3YSGwmNVRwHNVyaUhzWRYrCZzS2pcoxIji%2B8pulwZzeauGb2n2qVbCSQsD1J9NLbYZwkM62%2Bq8tQeWn%2FO3VGuAxrFvJIffEy%2BMMee1vmGBg%3D%3D'
train=pd.read_csv(url)
test.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

## getting to know the data

In [379]:
train.columns.values

array(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'], dtype=object)

In [380]:
combine =[train,test]

In [381]:
for dataset in combine:
    dataset['Title'] = dataset.Name.str.extract(' ([A-Za-z]+)\.', expand=False)

pd.crosstab(train['Title'], train['Sex'])

Sex       female  male
Title                 
Capt           0     1
Col            0     2
Countess       1     0
Don            0     1
Dr             1     6
Jonkheer       0     1
Lady           1     0
Major          0     2
Master         0    40
Miss         182     0
Mlle           2     0
Mme            1     0
Mr             0   517
Mrs          125     0
Ms             1     0
Rev            0     6
Sir            0     1

In [382]:
for dataset in combine:
    dataset['Title'] = dataset['Title'].replace(['Lady', 'Countess','Capt', 'Col',\
 	'Don', 'Dr', 'Major', 'Rev', 'Sir', 'Jonkheer', 'Dona'], 'Rare')

    dataset['Title'] = dataset['Title'].replace('Mlle', 'Miss')
    dataset['Title'] = dataset['Title'].replace('Ms', 'Miss')
    dataset['Title'] = dataset['Title'].replace('Mme', 'Mrs')
    
train[['Title', 'Survived']].groupby(['Title'], as_index=False).mean()

Title  Survived
0  Master  0.575000
1    Miss  0.702703
2      Mr  0.156673
3     Mrs  0.793651
4    Rare  0.347826

In [383]:
title_mapping = {"Mr": 1, "Miss": 2, "Mrs": 3, "Master": 4, "Rare": 5}
for dataset in combine:
    dataset['Title'] = dataset['Title'].map(title_mapping)
    dataset['Title'] = dataset['Title'].fillna(0)

train.head()


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  Title  
0      0         A/5 21171   7.2500   NaN        S      1  
1      0          PC 17599  71.2833   C85        C      3  
2      0  STON/O2. 3101282   7.9250   NaN        S      2  
3      0            113803  53.1000  C123        S      3  
4      0            373450   8.0500   NaN        S      1

In [384]:
train = train.drop(['Name', 'PassengerId'], axis=1)
test = test.drop(['Name'], axis=1)
combine = [train, test]


In [385]:
train.head()



Survived  Pclass     Sex   Age  SibSp  Parch            Ticket     Fare  \
0         0       3    male  22.0      1      0         A/5 21171   7.2500   
1         1       1  female  38.0      1      0          PC 17599  71.2833   
2         1       3  female  26.0      0      0  STON/O2. 3101282   7.9250   
3         1       1  female  35.0      1      0            113803  53.1000   
4         0       3    male  35.0      0      0            373450   8.0500   

  Cabin Embarked  Title  
0   NaN        S      1  
1   C85        C      3  
2   NaN        S      2  
3  C123        S      3  
4   NaN        S      1

## which features are categorical

Nominal data ---- They don't have any levels. we cannot do mean... it usually measured using percentage and frequency.
ordinal data ----  They measured in levels. (first,secon,third)






In [ ]:
nominal categorical data ---- survived,sex,embarked
ordinal categoricsl data ---- Pclass
numerical data ---- Age, Fare (discreet) --- sibsb,parch

## which features contain mixed data types

In [ ]:
mixed data --- ticket and cabin is alphanumeric

In [386]:
train.info()



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
Survived    891 non-null int64
Pclass      891 non-null int64
Sex         891 non-null object
Age         714 non-null float64
SibSp       891 non-null int64
Parch       891 non-null int64
Ticket      891 non-null object
Fare        891 non-null float64
Cabin       204 non-null object
Embarked    889 non-null object
Title       891 non-null int64
dtypes: float64(2), int64(5), object(4)
memory usage: 76.6+ KB


## which features may contain error or typos

In [ ]:
Name feature may contain errors or typos as there are several ways used to describe a name including titles, round brackets, and quotes used for alternative or short names.


## Which features contain blank, null or empty values?

In [ ]:
null values --- AGE,cabin

## what is the distribution of numerical features

This helps us determine, among other early insights, how representative is the training dataset of the actual problem domain.

1) Total samples are 891 or 40% of the actual number of passengers on board the Titanic (2,224).    
2) Survived is a categorical feature with 0 or 1 values.           
3) Around 38% samples survived representative of the actual survival rate at 32%.         
4) Most passengers (> 75%) did not travel with parents or children.                 
5) Nearly 30% of the passengers had siblings and/or spouse aboard.          
6) Fares varied significantly with few passengers (<1%) paying as high as $512.                 
7) Few elderly passengers (<1%) within age range 65-80.                     

In [387]:
train.describe()

Survived      Pclass         Age       SibSp       Parch        Fare  \
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000   
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208   
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429   
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000   
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400   
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200   
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000   
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200   

            Title  
count  891.000000  
mean     1.728395  
std      1.030039  
min      1.000000  
25%      1.000000  
50%      1.000000  
75%      2.000000  
max      5.000000

## distribution for categorical values 

In [388]:
train.describe(include=['O'])



Sex  Ticket        Cabin Embarked
count    891     891          204      889
unique     2     681          147        3
top     male  347082  C23 C25 C27        S
freq     577       7            4      644

## Assumtions based on data analysis

We arrive at following assumptions based on data analysis done so far. We may validate these assumptions further before taking appropriate actions.

## Correlating.

We want to know how well does each feature correlate with Survival. We want to do this early in our project and match these quick correlations with modelled correlations later in the project.

## Completing.

We may want to complete Age feature as it is definitely correlated to survival.
We may want to complete the Embarked feature as it may also correlate with survival or another important feature.
Correcting.

Ticket feature may be dropped from our analysis as it contains high ratio of duplicates (22%) and there may not be a correlation between Ticket and survival.
Cabin feature may be dropped as it is highly incomplete or contains many null values both in training and test dataset.
PassengerId may be dropped from training dataset as it does not contribute to survival.
Name feature is relatively non-standard, may not contribute directly to survival, so maybe dropped.

## Creating.

We may want to create a new feature called Family based on Parch and SibSp to get total count of family members on board.
We may want to engineer the Name feature to extract Title as a new feature.
We may want to create new feature for Age bands. This turns a continous numerical feature into an ordinal categorical feature.
We may also want to create a Fare range feature if it helps our analysis.

## Classifying.

We may also add to our assumptions based on the problem description noted earlier.

Women (Sex=female) were more likely to have survived.
Children (Age<?) were more likely to have survived.
The upper-class passengers (Pclass=1) were more likely to have survived.



## Analysing by pivoting feature 

To confirm some of our observations and assumptions, we can quickly analyze our feature correlations by pivoting features against each other. We can only do so at this stage for features which do not have any empty values. It also makes sense doing so only for features which are categorical (Sex), ordinal (Pclass) or discrete (SibSp, Parch) type.

. Pclass We observe significant correlation (>0.5) among Pclass=1 and Survived (classifying #3). We decide to include this feature in our model.        
. Sex We confirm the observation during problem definition that Sex=female had very high survival rate at 74% (classifying #1).         
. SibSp and Parch These features have zero correlation for certain values. It may be best to derive a feature or a set of features from these individual features (creating #1).     

In [389]:
train.groupby('Pclass').Survived.mean().sort_values(ascending=False)
train[['Pclass','Survived']].groupby('Pclass',as_index=False).mean().sort_values(by='Survived',ascending=False)

Pclass  Survived
0       1  0.629630
1       2  0.472826
2       3  0.242363

In [390]:
train.groupby('Sex').Survived.mean().sort_values(ascending=False)

train[['Sex','Survived']].groupby('Sex',as_index=False).mean().sort_values(by='Survived',ascending=False)

Sex  Survived
0  female  0.742038
1    male  0.188908

In [392]:
train[['SibSp','Survived']].groupby('SibSp',as_index=False).mean().sort_values(by='Survived',ascending=False)


SibSp  Survived
1      1  0.535885
2      2  0.464286
0      0  0.345395
3      3  0.250000
4      4  0.166667
5      5  0.000000
6      8  0.000000

In [391]:
train[['Parch','Survived']].groupby('Parch',as_index=False).mean().sort_values(by='Survived',ascending=False)

Parch  Survived
3      3  0.600000
1      1  0.550847
2      2  0.500000
0      0  0.343658
5      5  0.200000
4      4  0.000000
6      6  0.000000

In [393]:
train=train.drop(['Cabin','Ticket'],axis=1)
test=test.drop(['Cabin','Ticket'],axis=1)


In [394]:
test.head()



PassengerId  Pclass     Sex   Age  SibSp  Parch     Fare Embarked  Title
0          892       3    male  34.5      0      0   7.8292        Q      1
1          893       3  female  47.0      1      0   7.0000        S      3
2          894       2    male  62.0      0      0   9.6875        Q      1
3          895       3    male  27.0      0      0   8.6625        S      1
4          896       3  female  22.0      1      1  12.2875        S      3

In [52]:
## dropping name and passenger id from the training set 


In [ ]:
train=train.drop(['Name','PassengerId'],axis=1)
test=test.drop(['Name'],axis=1)



In [396]:
train.head()

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked  Title
0         0       3    male  22.0      1      0   7.2500        S      1
1         1       1  female  38.0      1      0  71.2833        C      3
2         1       3  female  26.0      0      0   7.9250        S      2
3         1       1  female  35.0      1      0  53.1000        S      3
4         0       3    male  35.0      0      0   8.0500        S      1

## Converting a categorical feature

Now we can convert features which contain strings to numerical values. This is required by most model algorithms. Doing so will also help us in achieving the feature completing goal.

Let us start by converting Sex feature to a new feature called Gender where female=1 and male=0.

In [398]:
train['Sex']=train.Sex.replace({'male':0,'female':1})

In [399]:
test['Sex']=test.Sex.replace({'male':0,'female':1})



In [400]:
train.head()
    

Survived  Pclass  Sex   Age  SibSp  Parch     Fare Embarked  Title
0         0       3    0  22.0      1      0   7.2500        S      1
1         1       1    1  38.0      1      0  71.2833        C      3
2         1       3    1  26.0      0      0   7.9250        S      2
3         1       1    1  35.0      1      0  53.1000        S      3
4         0       3    0  35.0      0      0   8.0500        S      1

In [401]:
test.head()

PassengerId  Pclass  Sex   Age  SibSp  Parch     Fare Embarked  Title
0          892       3    0  34.5      0      0   7.8292        Q      1
1          893       3    1  47.0      1      0   7.0000        S      3
2          894       2    0  62.0      0      0   9.6875        Q      1
3          895       3    0  27.0      0      0   8.6625        S      1
4          896       3    1  22.0      1      1  12.2875        S      3

## now we need to fill null values of age with the median or avg values

In [402]:
train['Age']=train['Age'].fillna(value=train.Age.mean()).astype(int)
test['Age']=test['Age'].fillna(value=test.Age.mean()).astype(int)


In [403]:
test.head()



PassengerId  Pclass  Sex  Age  SibSp  Parch     Fare Embarked  Title
0          892       3    0   34      0      0   7.8292        Q      1
1          893       3    1   47      1      0   7.0000        S      3
2          894       2    0   62      0      0   9.6875        Q      1
3          895       3    0   27      0      0   8.6625        S      1
4          896       3    1   22      1      1  12.2875        S      3

## let us replace age with ordinals

In [404]:
combine=[train,test]
for dataset in combine:    
    dataset.loc[ dataset['Age'] <= 16, 'Age'] = 0
    dataset.loc[(dataset['Age'] > 16) & (dataset['Age'] <= 32), 'Age'] = 1
    dataset.loc[(dataset['Age'] > 32) & (dataset['Age'] <= 48), 'Age'] = 2
    dataset.loc[(dataset['Age'] > 48) & (dataset['Age'] <= 64), 'Age'] = 3
    dataset.loc[ dataset['Age'] > 64, 'Age']


In [405]:
train.head()

Survived  Pclass  Sex  Age  SibSp  Parch     Fare Embarked  Title
0         0       3    0    1      1      0   7.2500        S      1
1         1       1    1    2      1      0  71.2833        C      3
2         1       3    1    1      0      0   7.9250        S      2
3         1       1    1    2      1      0  53.1000        S      3
4         0       3    0    2      0      0   8.0500        S      1

In [406]:
test.head()

PassengerId  Pclass  Sex  Age  SibSp  Parch     Fare Embarked  Title
0          892       3    0    2      0      0   7.8292        Q      1
1          893       3    1    2      1      0   7.0000        S      3
2          894       2    0    3      0      0   9.6875        Q      1
3          895       3    0    1      0      0   8.6625        S      1
4          896       3    1    1      1      1  12.2875        S      3

## create new feature combining existing feature

We can create a new feature for FamilySize which combines Parch and SibSp. This will enable us to drop Parch and SibSp from our datasets.

In [407]:
for dataset in combine:
    dataset['FamilySize'] = dataset['SibSp'] + dataset['Parch'] + 1


In [408]:
train[['FamilySize','Survived']].groupby('FamilySize',as_index=False).mean().sort_values(by='Survived',ascending=False)

FamilySize  Survived
3           4  0.724138
2           3  0.578431
1           2  0.552795
6           7  0.333333
0           1  0.303538
4           5  0.200000
5           6  0.136364
7           8  0.000000
8          11  0.000000

We can create another feature called IsAlone.

In [409]:
for dataset in combine:
    dataset['IsAlone'] = 0
    dataset.loc[dataset['FamilySize'] == 1, 'IsAlone'] = 1

train[['IsAlone', 'Survived']].groupby(['IsAlone'], as_index=False).mean()

IsAlone  Survived
0        0  0.505650
1        1  0.303538

Let us drop Parch, SibSp, and FamilySize features in favor of IsAlone.

In [410]:
train.drop(['Parch','SibSp','FamilySize'],axis=1,inplace=True)
test.drop(['Parch','SibSp','FamilySize'],axis=1,inplace=True)



In [411]:
train.head()


Survived  Pclass  Sex  Age     Fare Embarked  Title  IsAlone
0         0       3    0    1   7.2500        S      1        0
1         1       1    1    2  71.2833        C      3        0
2         1       3    1    1   7.9250        S      2        1
3         1       1    1    2  53.1000        S      3        0
4         0       3    0    2   8.0500        S      1        1

In [412]:
test.head()


PassengerId  Pclass  Sex  Age     Fare Embarked  Title  IsAlone
0          892       3    0    2   7.8292        Q      1        1
1          893       3    1    2   7.0000        S      3        0
2          894       2    0    3   9.6875        Q      1        1
3          895       3    0    1   8.6625        S      1        1
4          896       3    1    1  12.2875        S      3        0

We can also create an artificial feature combining Pclass and Age.

In [413]:
for dataset in combine:
    dataset['Age*Class'] = dataset.Age * dataset.Pclass

train.loc[:, ['Age*Class', 'Age', 'Pclass']].head(10)

Age*Class  Age  Pclass
0          3    1       3
1          2    2       1
2          3    1       3
3          2    2       1
4          6    2       3
5          3    1       3
6          3    3       1
7          0    0       3
8          3    1       3
9          0    0       2

## Completing a categorical feature

Embarked feature takes S, Q, C values based on port of embarkation. Our training dataset has two missing values. We simply fill these with the most common occurance.

In [414]:
freq_embarked=train.Embarked.mode()[0]
freq_embarked


'S'

In [415]:
train['Embarked']=train.Embarked.fillna(freq_embarked)

In [416]:
train[['Embarked','Survived']].groupby('Embarked',as_index=False).mean().sort_values(by='Survived',ascending=False)



Embarked  Survived
0        C  0.553571
1        Q  0.389610
2        S  0.339009

In [417]:
# converting embarked terms into 1,2,3

train['Embarked']=train['Embarked'].map({'S':0,'C':1,'Q':2}).astype(int)
test['Embarked']=test['Embarked'].map({'S':0,'C':1,'Q':2}).astype(int)

In [418]:
train.head()

Survived  Pclass  Sex  Age     Fare  Embarked  Title  IsAlone  Age*Class
0         0       3    0    1   7.2500         0      1        0          3
1         1       1    1    2  71.2833         1      3        0          2
2         1       3    1    1   7.9250         0      2        1          3
3         1       1    1    2  53.1000         0      3        0          2
4         0       3    0    2   8.0500         0      1        1          6

## Quick completing and converting a numeric feature

We can now complete the Fare feature for single missing value in test dataset using mode to get the value that occurs most frequently for this feature. We do this in a single line of code.

Note that we are not creating an intermediate new feature or doing any further analysis for correlation to guess missing feature as we are replacing only a single value. The completion goal achieves desired requirement for model algorithm to operate on non-null values.

We may also want round off the fare to two decimals as it represents currency.

In [419]:
fare_med=test['Fare'].median()
test['Fare']=test['Fare'].fillna(fare_med)


In [420]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 9 columns):
PassengerId    418 non-null int64
Pclass         418 non-null int64
Sex            418 non-null int64
Age            418 non-null int64
Fare           418 non-null float64
Embarked       418 non-null int64
Title          418 non-null int64
IsAlone        418 non-null int64
Age*Class      418 non-null int64
dtypes: float64(1), int64(8)
memory usage: 29.5 KB


In [421]:
train['FareBand'] = pd.qcut(train['Fare'], 4)
train[['FareBand', 'Survived']].groupby(['FareBand'], as_index=False).mean().sort_values(by='FareBand', ascending=True)

FareBand  Survived
0   (-0.001, 7.91]  0.197309
1   (7.91, 14.454]  0.303571
2   (14.454, 31.0]  0.454955
3  (31.0, 512.329]  0.581081

In [301]:
train['FareBand'] = pd.qcut(train['Fare'], 4)
train[['FareBand', 'Survived']].groupby(['FareBand'], as_index=False).mean().sort_values(by='FareBand', ascending=True)

FareBand  Survived
0   (-0.001, 7.91]  0.197309
1   (7.91, 14.454]  0.303571
2   (14.454, 31.0]  0.454955
3  (31.0, 512.329]  0.581081

Convert the Fare feature to ordinal values based on the FareBand.

In [422]:
for dataset in combine:
    dataset.loc[ dataset['Fare'] <= 7.91, 'Fare'] = 0
    dataset.loc[(dataset['Fare'] > 7.91) & (dataset['Fare'] <= 14.454), 'Fare'] = 1
    dataset.loc[(dataset['Fare'] > 14.454) & (dataset['Fare'] <= 31), 'Fare']   = 2
    dataset.loc[ dataset['Fare'] > 31, 'Fare'] = 3
    dataset['Fare'] = dataset['Fare'].astype(int)

train = train.drop(['FareBand'], axis=1)
combine = [train, test]
    
train.head(10)

Survived  Pclass  Sex  Age  Fare  Embarked  Title  IsAlone  Age*Class
0         0       3    0    1     0         0      1        0          3
1         1       1    1    2     3         1      3        0          2
2         1       3    1    1     1         0      2        1          3
3         1       1    1    2     3         0      3        0          2
4         0       3    0    2     1         0      1        1          6
5         0       3    0    1     1         2      1        1          3
6         0       1    0    3     3         0      1        1          3
7         0       3    0    0     2         0      4        0          0
8         1       3    1    1     1         0      3        0          3
9         1       2    1    0     2         1      3        0          0

In [423]:
test.head()

PassengerId  Pclass  Sex  Age  Fare  Embarked  Title  IsAlone  Age*Class
0          892       3    0    2     0         2      1        1          6
1          893       3    1    2     0         0      3        0          6
2          894       2    0    3     1         2      1        1          6
3          895       3    0    1     1         0      1        1          3
4          896       3    1    1     1         0      3        0          3

## Model, predict and solve


Now we are ready to train a model and predict the required solution. There are 60+ predictive modelling algorithms to choose from. We must understand the type of problem and solution requirement to narrow down to a select few models which we can evaluate. Our problem is a classification and regression problem. We want to identify relationship between output (Survived or not) with other variables or features (Gender, Age, Port...). We are also perfoming a category of machine learning which is called supervised learning as we are training our model with a given dataset. With these two criteria - Supervised Learning plus Classification and Regression, we can narrow down our choice of models to a few. These include:

Logistic Regression                                                 
KNN or k-Nearest Neighbors                                                     
Support Vector Machines                                                 
Naive Bayes classifier                                               
Decision Tree                                                   
Random Forrest                                       
Perceptron                                                    
Artificial neural network                                     
RVM or Relevance Vector Machine                                                  

In [424]:
X_train=train.drop('Survived',axis=1)
Y_train=train['Survived']
X_test=test.drop('PassengerId',axis=1).copy()
X_train.shape,Y_train.shape,X_test.shape


((891, 8), (891,), (418, 8))

Logistic Regression is a useful model to run early in the workflow. Logistic regression measures the relationship between the categorical dependent variable (feature) and one or more independent variables (features) by estimating probabilities using a logistic function, which is the cumulative logistic distribution. Reference Wikipedia.

Note the confidence score generated by the model based on our training dataset.

In [430]:
# logistic regression
from sklearn.linear_model import LogisticRegression
from sklearn.cross_validation import train_test_split

lm=LogisticRegression()
model=lm.fit(X_train,Y_train)
predict=lm.predict(X_test)